File: modules/backlogs/spec/controllers/projects/settings/backlogs_controller_spec.rb
# frozen_string_literal: true

#-- copyright
# OpenProject is an open source project management software.
# Copyright (C) the OpenProject GmbH
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License version 3.
#
# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
# Copyright (C) 2006-2013 Jean-Philippe Lang
# Copyright (C) 2010-2013 the ChiliProject Team
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
# See COPYRIGHT and LICENSE files for more details.
#++

require "spec_helper"

RSpec.describe Projects::Settings::BacklogsController do
  shared_let(:user) { create(:admin) }

  current_user { user }

  before do
    visible_relation = instance_double(ActiveRecord::Relation)
    allow(Project).to receive(:visible).and_return(visible_relation)
    allow(visible_relation).to receive(:find).with(project.identifier).and_return(project)
  end

  describe "PATCH #update" do
    let(:project) { build_stubbed(:project) }
    let(:service_result) { ServiceResult.success(result: project) }
    let(:update_service) { instance_double(Projects::UpdateService, call: service_result) }

    before do
      allow(Projects::UpdateService)
        .to receive(:new)
        .with(model: project, user:, contract_class: Backlogs::Projects::BacklogsTypesAndStatusesContract)
        .and_return(update_service)

      patch :update, params: { project_id: project.identifier, project: project_params }
    end

    context "when service call succeeds" do
      let(:project_params) do
        {
          backlog_excluded_type_ids: ["3"],
          done_status_ids: %w[1 2],
          name: "must_be_ignored"
        }
      end

      it "updates backlogs type and status settings and redirects to show", :aggregate_failures do
        expect(update_service).to have_received(:call).with(
          ActionController::Parameters.new(
            "done_status_ids" => %w[1 2],
            "backlog_excluded_type_ids" => ["3"]
          ).permit!
        )

        expect(response).to redirect_to(project_settings_backlogs_path(project))
        expect(flash[:notice]).to include I18n.t(:notice_successful_update)
      end
    end

    context "when service call fails" do
      let(:service_result) { ServiceResult.failure(result: project, message: "invalid setting") }
      let(:project_params) do
        {
          backlog_excluded_type_ids: [],
          done_status_ids: ["999"]
        }
      end

      it "renders show with an error", :aggregate_failures do
        expect(response).to have_http_status(:unprocessable_entity)
        expect(response).to render_template("projects/settings/backlogs/show")
        expect(flash[:error]).to eq I18n.t(:notice_unsuccessful_update_with_reason, reason: "invalid setting")
      end
    end

    context "when only hidden empty value is sent for an empty done_status_ids multi-select" do
      let(:project_params) do
        {
          done_status_ids: [""],
          backlog_excluded_type_ids: ["3"]
        }
      end

      it "passes through the Rails hidden-field payload", :aggregate_failures do
        expect(update_service).to have_received(:call).with(
          ActionController::Parameters.new(
            "done_status_ids" => [""],
            "backlog_excluded_type_ids" => ["3"]
          ).permit!
        )

        expect(response).to redirect_to(project_settings_backlogs_path(project))
        expect(flash[:notice]).to include I18n.t(:notice_successful_update)
      end
    end

    context "when duplicate ids are sent" do
      let(:project_params) do
        {
          backlog_excluded_type_ids: %w[3 3],
          done_status_ids: %w[2 2]
        }
      end

      it "deduplicates them before passing to the service", :aggregate_failures do
        expect(update_service).to have_received(:call).with(
          ActionController::Parameters.new(
            "done_status_ids" => ["2"],
            "backlog_excluded_type_ids" => ["3"]
          ).permit!
        )

        expect(response).to redirect_to(project_settings_backlogs_path(project))
        expect(flash[:notice]).to include I18n.t(:notice_successful_update)
      end
    end
  end
end

File: frontend/src/stimulus/controllers/dynamic/backlogs.controller.spec.ts
//-- copyright
// OpenProject is an open source project management software.
// Copyright (C) the OpenProject GmbH
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License version 3.
//
// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
// Copyright (C) 2006-2013 Jean-Philippe Lang
// Copyright (C) 2010-2013 the ChiliProject Team
//
// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// See COPYRIGHT and LICENSE files for more details.
//++

import { waitFor } from '@testing-library/dom';
import { Subject } from 'rxjs';
import { vi, type Mock } from 'vitest';

import { setupStimulusTest, type StimulusTestContext } from 'core-stimulus/test-helpers';
import type BacklogsControllerType from './backlogs.controller';

interface HalEvent {
  eventType:string;
}

describe('Backlogs controller', () => {
  let ctx:StimulusTestContext;
  let BacklogsController:typeof BacklogsControllerType;
  let events$:Subject<HalEvent[]>;
  let aggregated$:Mock;
  let reload:Mock;
  let originalOpenProject:typeof window.OpenProject;

  const pluginContext = () => ({
    services: { halEvents: { aggregated$ } },
  });

  beforeAll(async () => {
    ({ default: BacklogsController } = await import('./backlogs.controller'));
  });

  beforeEach(async () => {
    events$ = new Subject<HalEvent[]>();
    aggregated$ = vi.fn(() => events$);
    reload = vi.fn(() => Promise.resolve());
    originalOpenProject = window.OpenProject;
    window.OpenProject = {
      getPluginContext: () => Promise.resolve(pluginContext()),
    } as unknown as typeof window.OpenProject;

    ctx = await setupStimulusTest({
      controllers: { backlogs: BacklogsController },
    });
  });

  afterEach(() => {
    ctx.dispose();
    window.OpenProject = originalOpenProject;
    vi.restoreAllMocks();
  });

  async function renderBacklogs() {
    await ctx.mount(`
      <div data-controller="backlogs">
        <turbo-frame id="backlogs_container"></turbo-frame>
      </div>
    `);
    const frame = ctx.container.querySelector('#backlogs_container')!;
    (frame as unknown as { reload:Mock }).reload = reload;
    return ctx.getController<BacklogsControllerType>('backlogs');
  }

  it('subscribes to aggregated work package events once connected', async () => {
    await renderBacklogs();

    await waitFor(() => {
      expect(aggregated$).toHaveBeenCalledWith('WorkPackage');
    });
  });

  it('reloads the list frame on an updated event', async () => {
    await renderBacklogs();
    await waitFor(() => { expect(aggregated$).toHaveBeenCalled(); });

    events$.next([{ eventType: 'updated' }]);

    await waitFor(() => {
      expect(reload).toHaveBeenCalled();
    });
  });

  it('ignores event batches without an update', async () => {
    await renderBacklogs();
    await waitFor(() => { expect(aggregated$).toHaveBeenCalled(); });

    events$.next([{ eventType: 'created' }]);
    await ctx.nextFrame();

    expect(reload).not.toHaveBeenCalled();
  });

  it('stops reloading once disconnected', async () => {
    await renderBacklogs();
    await waitFor(() => { expect(aggregated$).toHaveBeenCalled(); });

    const root = ctx.container.querySelector('[data-controller="backlogs"]')!;
    root.remove();
    await ctx.nextFrame();

    events$.next([{ eventType: 'updated' }]);
    await ctx.nextFrame();

    expect(reload).not.toHaveBeenCalled();
  });

  it('does not subscribe when disconnected before the context resolves', async () => {
    let resolveContext!:(context:unknown) => void;
    window.OpenProject = {
      getPluginContext: () => new Promise((resolve) => { resolveContext = resolve; }),
    } as unknown as typeof window.OpenProject;

    await renderBacklogs();
    const root = ctx.container.querySelector('[data-controller="backlogs"]')!;

    root.remove();
    await ctx.nextFrame();

    resolveContext(pluginContext());
    await ctx.nextFrame();

    expect(aggregated$).not.toHaveBeenCalled();
  });
});
